services/worker-manager/src/providers/aws.js

```diff
@@ -121,6 +135,151 @@ class AwsProvider extends Provider {
 
     //todo: do something to enable logging to stackdriver
   }
+
+  async provision({workerPool}) {
+    const {workerPoolId} = workerPool;
+
+    if (!workerPool.providerData[this.providerId] || workerPool.providerData[this.providerId].running === undefined) {
+      await workerPool.modify(wt => {
+        wt.providerData[this.providerId] = wt.providerData[this.providerId] || {};
+        wt.providerData[this.providerId].running = wt.providerData[this.providerId].running || 0; // why zero?
+      });
+    }
+
+    const toSpawn = await this.estimator.simple({
+      workerPoolId,
+      ...workerPool.config,
+      running: workerPool.providerData[this.providerId].running,
+    });
+
+    let spawned;
+
+    try {
+      spawned = await this.ec2.runInstances({
+        IamInstanceProfile: {
+          Arn: this.instanceProfile.Arn,
+          Name: this.instanceProfile.InstanceProfileName,
+        },
+        MaxCount: toSpawn,
+        MinCount: toSpawn,
+        ImageId: workerPool.config.imageId,
+        InstanceType: workerPool.config.instanceType,
+        Placement: {
+          AvailabilityZone: this.region,
+        },
+        TagSpecifications: {
+          ResourceType: 'instance',
+          Tags: [{
+            Key: 'Provider',
+            Value: `wm-${this.providerId}`,
+          }, {
+            Key: 'Owner',
+            Value: workerPool.owner,
+          }, {
+            Key: 'Project',
+            Value: this.project,
+          }],
+        },
+      });
+    } catch (e) {
+      await workerPool.reportError({
+        kind: 'creation-error',
+        title: 'Instance Creation Error',
+        description: e.message, // TODO: Make sure we clear exposing this with security folks
+        notify: this.notify,
+        WorkerPoolError: this.WorkerPoolError,
+      });
+    }
+
+    Promise.all(spawned.Instances.map(i => {
+      return this.Worker.create({
+        workerPoolId,
+        providerId: this.providerId,
+        workerGroup: this.providerId,
+        workerId: i.InstanceId,
+        created: new Date(),
+        expires: taskcluster.fromNow('1 week'),
+        state: this.Worker.states.REQUESTED,
+        providerData: { // what do we want to remember, anyways? 🤔
```

Comment:
@imbstack @djmitche I have one more question^  I guess this is information that either w-m or other services might need about the worker? Any use cases for this data? Or suggestions what to put in there? (maybe I'd just leave for now and add things as the need arises?)

Reply:
It is only for w-m to use itself. It doesn't get returned by any api calls. I use it to remember things about a worker I'll need later to confirm the request was completed, etc.